File: cpp/src/fdx/parser.cpp
// This file is part of an MIT-licensed project: see LICENSE file or README.md
// for details. Copyright (c) 2024 Ian Thomas

#include "screenplay_tools/fdx/parser.h"
#include "xml_helper.h"

namespace ScreenplayTools {
namespace FDX {

Parser::Parser() {}

Script Parser::Parse(const std::string &xmlContent) {
  Script script;

  // Simple preprocessing to verify empty or too short
  if (xmlContent.empty())
    return script;

  try {
    XMLElement root = XMLHelper::Parse(xmlContent);

    // Navigate to Content -> Paragraph
    // root should be FinalDraft (or whatever XMLHelper returned as top level)
    // XMLHelper::Parse might return the root node found.

    const XMLElement *finalDraft = nullptr;
    if (root.name == "FinalDraft") {
      finalDraft = &root;
    }

    if (!finalDraft)
      return script;

    const XMLElement *content = nullptr;
    for (const auto &child : finalDraft->children) {
      if (child.name == "Content") {
        content = &child;
        break;
      }
    }

    if (!content)
      return script;

    for (const auto &p : content->children) {
      if (p.name != "Paragraph")
        continue;

      std::string type = "Action"; // default
      if (p.attributes.count("Type")) {
        type = p.attributes.at("Type");
      }

      // Extract text
      // FDX: <Paragraph> <Text>Foo</Text> <Text>Bar</Text> </Paragraph>
      std::string text = "";
      for (const auto &txt : p.children) {
        if (txt.name == "Text") {
          text += txt.text;
        }
      }

      // Normalize text and map types
      // Clean up text
      // Assuming simplified text extraction from XMLHelper for now.
      // If XMLHelper returned child text nodes as part of children or text
      // property... In my XMLHelper, text directly in an element is in .text,
      // children are in .children. <Text>content</Text> -> name="Text",
      // text="content"

      // Element creation
      if (type == "Scene Heading" || type == "Scene Heading (Top of Page)" ||
          type == "Shot") {
        script.addElement(std::make_shared<SceneHeading>(text));
      } else if (type == "Action" || type == "General") {
        script.addElement(std::make_shared<Action>(text));
      } else if (type == "Character") {
        std::string name = text;
        std::string extension = "";

        // Parse NAME (EXT)
        // Trim
        while (!name.empty() && std::isspace(name.back()))
          name.pop_back();
        while (!name.empty() && std::isspace(name.front()))
          name.erase(0, 1);

        if (!name.empty() && name.back() == ')') {
          size_t openParen = name.rfind("(");
          if (openParen != std::string::npos && openParen > 0) {
            extension =
                name.substr(openParen + 1, name.length() - openParen - 2);
            name = name.substr(0, openParen);
            // Trim name again
            while (!name.empty() && std::isspace(name.back()))
              name.pop_back();
          }
        }

        std::optional<std::string> extOpt = std::nullopt;
        if (!extension.empty())
          extOpt = extension;

        script.addElement(std::make_shared<Character>(name, extOpt));
      } else if (type == "Dialogue") {
        script.addElement(std::make_shared<Dialogue>(text));
      } else if (type == "Parenthetical") {
        std::string pText = text;
        // Trim
        while (!pText.empty() && std::isspace(pText.back()))
          pText.pop_back();
        while (!pText.empty() && std::isspace(pText.front()))
          pText.erase(0, 1);

        if (pText.size() >= 2 && pText.front() == '(' && pText.back() == ')') {
          pText = pText.substr(1, pText.length() - 2);
          // Trim
          while (!pText.empty() && std::isspace(pText.back()))
            pText.pop_back();
          while (!pText.empty() && std::isspace(pText.front()))
            pText.erase(0, 1);
        }
        script.addElement(std::make_shared<Parenthetical>(pText));
      } else if (type == "Transition") {
        script.addElement(std::make_shared<Transition>(text));
      } else {
        script.addElement(std::make_shared<Action>(text));
      }
    }
  } catch (...) {
    // Silent failure return empty script? Or partial?
  }

  return script;
}

} // namespace FDX
} // namespace ScreenplayTools

File: javascript/src/fdx/parser.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

import { XMLParser } from 'fast-xml-parser';
import { Script, Action, SceneHeading, Character, Dialogue, Parenthetical, Transition } from '../screenplay.js';

export class FDXParser {
    constructor() {
    }

    /**
     * Parse an FDX XML string into a Script object.
     * @param {string} xmlContent - The FDX XML string.
     * @returns {Script} The parsed script.
     */
    parse(xmlContent) {
        let script = new Script();

        // fast-xml-parser options
        const options = {
            ignoreAttributes: false,
            attributeNamePrefix: "@_",
            textNodeName: "Text"
        };
        const parser = new XMLParser(options);

        // Pre-processing
        // Remove XML declaration if present to avoid any parsing issues similar to C# (though JS usually lenient)
        xmlContent = xmlContent.trim();
        // fast-xml-parser handles declarations fine usually, but let's be safe if needed.
        // For now, raw pass.

        let jsonObj;
        try {
            jsonObj = parser.parse(xmlContent);
        } catch (error) {
            throw new Error(`Failed to parse XML: ${error.message}`);
        }

        if (!jsonObj.FinalDraft || !jsonObj.FinalDraft.Content || !jsonObj.FinalDraft.Content.Paragraph) {
            // Empty script or invalid FDX structure
            return script;
        }

        // Paragraph can be an array or a single object if only one paragraph
        let paragraphs = jsonObj.FinalDraft.Content.Paragraph;
        if (!Array.isArray(paragraphs)) {
            paragraphs = [paragraphs];
        }

        for (const p of paragraphs) {
            const type = p["@_Type"] || "Action";
            // Text can be complex in FDX (mixed content), but fast-xml-parser might map it to 'Text' property if configured?

            let text = "";
            if (p.Text) {
                const textNodes = Array.isArray(p.Text) ? p.Text : [p.Text];
                text = textNodes.map(node => {
                    if (typeof node === 'string') return node;
                    // If textNodeName is "Text", the content is in node.Text
                    if (node && node.Text) return node.Text;
                    // If default #text processing or mixed content
                    if (node && node["#text"]) return node["#text"];
                    return "";
                }).join("");
            }

            switch (type) {
                case "Scene Heading":
                case "Scene Heading (Top of Page)": // FDX Sometimes has this
                case "Shot":
                    script.addElement(new SceneHeading(text));
                    break;
                case "Action":
                case "General":
                    script.addElement(new Action(text));
                    break;
                case "Character":
                    // Parse NAME (EXT)
                    let charText = text.trim();
                    let extension = null;
                    if (charText.endsWith(")")) {
                        let openParen = charText.lastIndexOf("(");
                        if (openParen > 0) {
                            extension = charText.substring(openParen + 1, charText.length - 1).trim(); // remove closing paren
                            charText = charText.substring(0, openParen).trim();
                        }
                    }
                    script.addElement(new Character(charText, extension));
                    break;
                case "Dialogue":
                    script.addElement(new Dialogue(text));
                    break;
                case "Parenthetical":
                    // Strip parens
                    let pText = text.trim();
                    if (pText.startsWith("(") && pText.endsWith(")")) {
                        pText = pText.substring(1, pText.length - 1).trim();
                    }
                    script.addElement(new Parenthetical(pText));
                    break;
                case "Transition":
                    script.addElement(new Transition(text));
                    break;
                default:
                    // Treat unknown as Action
                    script.addElement(new Action(text));
                    break;
            }
        }

        return script;
    }
}
